src/com/sun/jna/internal/Cleaner.java

```diff
@@ -35,84 +36,80 @@
  * objects. It replaces the {@code Object#finalize} based resource deallocation
  * that is deprecated for removal from the JDK.
  *
- * <p><strong>This class is intented to be used only be JNA itself.</strong></p>
+ * <p><strong>This class is intended to be used only be JNA itself.</strong></p>
  */
 public class Cleaner {
     private static final Cleaner INSTANCE = new Cleaner();
+    private static final Logger logger = Logger.getLogger(Cleaner.class.getName());
+    private static final long CLEANER_LINGER_TIME = TimeUnit.SECONDS.toMillis(30);
 
     public static Cleaner getCleaner() {
         return INSTANCE;
     }
 
     private final ReferenceQueue<Object> referenceQueue;
-    private Thread cleanerThread;
+    private boolean cleanerRunning;
     private CleanerRef firstCleanable;
 
     private Cleaner() {
         referenceQueue = new ReferenceQueue<>();
     }
 
-    public synchronized Cleanable register(Object obj, Runnable cleanupTask) {
-        // The important side effect is the PhantomReference, that is yielded
-        // after the referent is GCed
-        return add(new CleanerRef(this, obj, referenceQueue, cleanupTask));
-    }
+    public synchronized Cleanable register(final Object obj, final Runnable cleanupTask) {
+        // The important side effect is the PhantomReference, that is yielded after the referent is GCed
+        final CleanerRef ref = new CleanerRef(obj, referenceQueue, cleanupTask);
 
-    private synchronized CleanerRef add(CleanerRef ref) {
-        synchronized (referenceQueue) {
-            if (firstCleanable == null) {
-                firstCleanable = ref;
-            } else {
-                ref.setNext(firstCleanable);
-                firstCleanable.setPrevious(ref);
-                firstCleanable = ref;
-            }
-            if (cleanerThread == null) {
-                Logger.getLogger(Cleaner.class.getName()).log(Level.FINE, "Starting CleanerThread");
-                cleanerThread = new CleanerThread();
-                cleanerThread.start();
-            }
-            return ref;
+        if (firstCleanable != null) {
+            ref.setNext(firstCleanable);
+            firstCleanable.setPrevious(ref);
+        }
+        firstCleanable = ref;
+
+        if (!cleanerRunning) {
+            logger.log(Level.FINE, "Starting CleanerThread");
+            Thread cleanerThread = new CleanerThread();
+            cleanerThread.start();
+            cleanerRunning = true;
         }
+
+        return ref;
```

Comment:
At this point we need the equivalent of `Reference.reachabilityFence` on `obj`. If the caller does not retain a strong reference, we need to ensure, that the reference is kept at least until the reference cleaner is completely enqueued. As observed in the last iteration early GC can happen: https://github.com/java-native-access/jna/pull/1684#issuecomment-3324165057.

In the comment I suggested to use an empty `sychronized` block.